src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,840 @@
+                         int32_t w, int32_t h,
+                         GError **err) {
+  struct dicom_ops_data *data = osr->data;
+  struct dicom_level *l = (struct dicom_level *) level;
+
+  g_debug("paint_region: x = %" PRId64 ", y = %" PRId64 ", w = %d, h = %d",
+          x, y, w, h);
+  g_debug("paint_region level:");
+  print_level(l);
+
+  return _openslide_grid_paint_region(l->grid, cr, data,
+                                      x / l->base.downsample,
+                                      y / l->base.downsample,
+                                      level, w, h,
+                                      err);
+}
+
+static const struct _openslide_ops dicom_ops = {
+  .paint_region = paint_region,
+  .destroy = destroy,
+};
+
+static bool is_type(struct dicom_file *f, const struct allowed_types *types) {
+  // ImageType must be one of the combinations we accept
+  for (int i = 0; i < types->n_types; i++) {
+    bool found_difference = false;
+
+    for (int j = 0; j < 4; j++) {
+      const char *type;
+      if (!get_tag_str(f->metadata, "ImageType", j, &type)) {
+        return false;
+      }
+
+      if (!g_str_equal(type, types->types[i][j])) {
+        found_difference = true;
+        break;
+      }
+    }
+
+    if (!found_difference) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+static bool associated_get_argb_data(struct _openslide_associated_image *_img,
+                                     uint32_t *dest,
+                                     GError **err) {
+  struct associated *a = (struct associated *) _img;
+
+  DcmError *dcm_error = NULL;
+  DcmFrame *frame = dcm_filehandle_read_frame(&dcm_error,
+                                              a->file->filehandle,
+                                              a->file->metadata,
+                                              a->file->bot,
+                                              1);
+  if (frame == NULL) {
+    dicom_propagate_error(err, dcm_error);
+    return false;
+  }
+
+  print_frame(frame);
+
+  return _openslide_jpeg_decode_buffer(dcm_frame_get_value(frame),
+                                       dcm_frame_get_length(frame),
+                                       dest,
+                                       a->base.w, a->base.h,
+                                       err);
+}
+
+static void associated_destroy(struct associated *a) {
+  dicom_file_destroy(a->file);
+  g_slice_free(struct associated, a);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(associated, associated_destroy)
+
+static void openslide_associated_destroy(struct _openslide_associated_image *_img) {
+  struct associated *a = (struct associated *) _img;
+  associated_destroy(a);
+}
+
+static const struct _openslide_associated_image_ops dicom_associated_ops = {
+  .get_argb_data = associated_get_argb_data,
+  .destroy = openslide_associated_destroy
+};
+
+static bool read_whole_file(struct dicom_file *f) {
+  // try to read the rest of the dicom file
+  if (!f->metadata) {
```

Comment:
Ideally we'd get to the point where this function is called at most once per `dicom_file` (or inlined into `dicom_open`) and we don't need the conditionals.